File: source/SSSR.py
from Graph import Graph
from typing import List, Final, Callable, Dict, Set, Tuple

class SSSR:
    ' A class encapsulating the functionality to find the smallest set of smallest rings in a graph. */'
    
    @staticmethod
    def getRings(graph: 'Graph', experimental: bool=False) -> List[List[float]]:
        'Returns an array containing arrays, each representing a ring from the smallest set of smallest rings in the graph.'
        adjacencyMatrix = graph.getComponentsAdjacencyMatrix()
        if len(adjacencyMatrix) == 0:
            return None
        connectedComponents = Graph.getConnectedComponents(adjacencyMatrix)
        for connectedComponent in connectedComponents:
            ccAdjacencyMatrix = graph.getSubgraphAdjacencyMatrix(*connectedComponent)
            length = len(ccAdjacencyMatrix)
            arrRingCount = [0] * len(ccAdjacencyMatrix)
            arrBondCount = [sum(row) for row in ccAdjacencyMatrix]
            nEdges = sum(elem for row in ccAdjacencyMatrix for elem in row)
            nSssr = nEdges - length + 1
            allThree: bool = all(elem == 3 for elem in arrBondCount) 
            
            if allThree:
                nSssr = 2.0 + nEdges - length
            if type(nSssr) is int and nSssr == 1:
                rings.append([*connectedComponent])
                continue
            if experimental:
                nSssr = 999
            d, pe, pe_prime = SSSR.getPathIncludedDistanceMatrices(ccAdjacencyMatrix).values()
            c = SSSR.getRingCandidates(d, pe, pe_prime)
            sssr = SSSR.getSSSR(c, d, ccAdjacencyMatrix, pe, pe_prime, arrBondCount, arrRingCount, nSssr)
            rings = [[connectedComponent[val] for val in sssr_elem] for sssr_elem in sssr ]
        return rings
    
    @staticmethod
    def matrixToString(matrix: List[List[float]]) -> str:
        'Creates a printable string from a matrix (2D array).'
        return '\n'.join(' '.join(f'{elem}' for elem in row) for row in matrix)

    @staticmethod
    def __initialize_matrices(length: int):
        'private method for getPathIncludedDistanceMatrices'
        d = [[float('inf')] * length for _ in range(length)]
        pe = [[[[] for _ in range(length)] for _ in range(length)] for _ in range(length)]
        pe_prime = [[[[] for _ in range(length)] for _ in range(length)] for _ in range(length)]
        return d, pe, pe_prime

    @staticmethod
    def __update_matrices(d: List[List[float]], pe: List[List[List[float]]],\
        pe_prime: List[List[List[float]]], i: int, j: int, k: int) -> None:
        'private method for getPathIncludedDistanceMatrices'
        previous_path_length = d[i][j]
        new_path_length = d[i][k] + d[k][j]
        if previous_path_length > new_path_length:
            d[i][j] = new_path_length
            pe[i][j] = [[]]
            pe[i][j][0] = pe[i][k][0] + pe[k][j][0]
            if previous_path_length == new_path_length + 1:
                pe_prime[i][j] = [[len(pe[i][j])]]
                for l in range(len(pe[i][j])):
                    pe_prime[i][j][l] = [[len(pe[i][j][l])]]
                    for m in range(len(pe[i][j][l])):
                        pe_prime[i][j][l][m] = [[len(pe[i][j][l][m])]]
                        for n in range(len(pe[i][j][l][m])):
                            pe_prime[i][j][l][m][n] = [pe[i][j][l][m][0], pe[i][j][l][m][1]]
            else:
                pe_prime[i][j] = [[]]
        elif previous_path_length == new_path_length:
            if pe[i][k] and pe[k][j]:
                if pe[i][j]:
                    pe[i][j].append(pe[i][k][0] + pe[k][j][0])
                else:
                    pe[i][j][0] = pe[i][k][0] + pe[k][j][0]
        elif previous_path_length == new_path_length - 1:
            if pe_prime[i][j]:
                pe_prime[i][j].append(pe[i][k][0] + pe[k][j][0])
            else:
                pe_prime[i][j][0] = pe[i][k][0] + pe[k][j][0]

    @staticmethod
    def getPathIncludedDistanceMatrices(adjacencyMatrix: List[List[float]]) -> dict:
        length = len(adjacencyMatrix)
        d, pe, pe_prime = SSSR.__initialize_matrices(length)
        for i in range(length):
            for j in range(length):
                d[i][j] = adjacencyMatrix[i][j] if i == j or adjacencyMatrix[i][j] == 1 else float('inf')
                if d[i][j] == 1:
                    pe[i][j] = [[[i, j]]]

        for k in range(length):
            for i in range(length):
                for j in range(length):
                    SSSR.__update_matrices(d, pe, pe_prime, i, j, k)
        return {'d': d, 'pe': pe, 'pe_prime': pe_prime}

    @staticmethod
    def getRingCandidates(d: List[List[float]], pe: List[List[float]],\
        pe_prime: List[List[float]]) -> List[List[float]]:
        'Get the ring candidates from the path-included distance matrices.'
        length = len(d)
        candidates = []
        c = 0
        for i in range(length):
            for j in range(length):
                if not (d[i][j] == 0 or (len(pe[i][j]) == 1 and pe_prime[i][j] == 0)):
                    c = 2 * (d[i][j] + 0.5) if len(pe_prime[i][j]) else 2 * d[i][j]
                    if c is not None:
                        candidates.append([c, pe[i][j], pe_prime[i][j]])
        candidates.sort(key=lambda x: x[0])
        return candidates

    @staticmethod
    def getSSSR(c: List[List[float]], d: List[List[float]], adjacencyMatrix: List[List[float]],
        pe: List[List[float]], pe_prime: List[List[float]], arrBondCount: List, arrRingCount: List,
        nsssr: float) -> List[Set]:
        'Searches the candidates for the smallest set of smallest rings.'
        c_sssr, all_bonds = [], []
        for i in range(len(c)):
            if c[i][0] % 2 != 0:
                c_sssr, all_bonds = SSSR.__process_odd_case(c[i],\
                    c_sssr, all_bonds, adjacencyMatrix, nsssr)
            else:
                c_sssr, all_bonds = SSSR.__process_even_case(c[i],\
                    c_sssr, all_bonds, adjacencyMatrix, nsssr)
        return c_sssr

    @staticmethod
    def __process_odd_case(data: List[float], c_sssr: List[Set], all_bonds: List[Set],
                        adjacencyMatrix: List[List[float]], nsssr: float) -> Tuple[List[Set], List[Set]]:
        'private method for getSSSR'
        for j in range(len(data[2])):
            bonds = data[1][0] + data[2][j]
            bonds = SSSR.__flatten_nested_lists(bonds)
            atoms = SSSR.bondsToAtoms(bonds)
            if SSSR.bondsToAtoms(atoms, adjacencyMatrix) == len(atoms) and \
                    not SSSR.pathSetsContain(c_sssr, atoms, bonds, all_bonds):
                c_sssr.append(atoms)
                all_bonds += bonds
                
            if len(c_sssr) > nsssr:
                return c_sssr, all_bonds
                
        return c_sssr, all_bonds

    @staticmethod
    def __process_even_case(data: List[float], c_sssr: List[Set], all_bonds: List[Set],
                        adjacencyMatrix: List[List[float]], nsssr: float) -> Tuple[List[Set], List[Set]]:
        'private method for getSSSR'
        for j in range(len(data[1]) - 1):
            bonds = data[1][j] + data[1][j + 1]
            bonds = SSSR.__flatten_nested_lists(bonds)
            atoms = SSSR.bonds_to_atoms(bonds)
            if SSSR.get_bond_count(atoms, adjacencyMatrix) == len(atoms) and \
                    not SSSR.pathSetsContain(c_sssr, atoms, bonds, all_bonds):
                c_sssr.append(atoms)
                all_bonds += bonds
                
            if len(c_sssr) > nsssr:
                return c_sssr, all_bonds
            
        return c_sssr, all_bonds

    @staticmethod
    def __flatten_nested_lists(lst: List[List]) -> List:
        'private method for getSSSR'
        return [item for sublist in lst for item in sublist]

    @staticmethod
    def getEdgeCount(adjacencyMatrix: List[List[int]]) -> int:
        'Returns the number of edges in a graph defined by an adjacency matrix.'
        return sum(row.count(1) for row in adjacencyMatrix)
    
    @staticmethod
    def getEdgeList(adjacencyMatrix: List[List[int]]) -> List[List[int]]:
        'Returns an edge list constructed form an adjacency matrix.'
        length = range(len(adjacencyMatrix))
        return [[i, j] for i in length for j in length if adjacencyMatrix[i][j] == 1]
    
    @staticmethod
    def bondsToAtoms(bonds: List[int]) -> Set[int]:
        'Return a set of vertex indices contained in an array of bonds.'
        atoms = set()
        i = len(bonds)
        while i > 0:
            i -= 1
            atoms.add(bonds[i][0])
            atoms.add(bonds[i][1])
        return atoms
    
    @staticmethod
    def bondsToAtoms(bonds: List[int]) -> Set[int]:
        'Return a set of vertex indices contained in an array of bonds.'
        return {bond[0] for bond in bonds} | {bond[1] for bond in bonds}
    
    @staticmethod
    def getBondCount(atoms: Set[int], adjacency_matrix: List[List[int]]) -> int:
        'Returns the number of bonds within a set of atoms.'
        return sum(adjacency_matrix[u][v] for u in atoms for v in atoms if u == v) // 2    
    
    @staticmethod
    def pathSetsContain(path_sets: Set, path_set: Set[int], bonds: List[List[int]],\
        all_bonds: List[List[int]], arr_bond_count: List, arr_ring_count: List) -> bool:
        'Checks whether or not a given path already exists in an array of paths.'
        for existing_set in path_sets:
            if path_set.issuperset(existing_set):
                return True
            if len(existing_set) != len(path_set):
                continue
            if existing_set == path_set:
                return True
    
        count = 0
        all_contained = False
        for bond in bonds:
            for all_bond in all_bonds:
                if (bond[0] == all_bond[0] and bond[1] == all_bond[1]) or \
                   (bond[1] == all_bond[0] and bond[0] == all_bond[1]):
                    count += 1
            if count == len(bonds):
                all_contained = True
    
        special_case = False
        if all_contained:
            for element in path_set:
                if arr_ring_count[element] < arr_bond_count[element]:
                    special_case = True
                    break
    
        if all_contained and not special_case:
            return True
        for element in path_set:
            arr_ring_count[element] += 1
        return False
    
    @staticmethod
    def areSetsEqual(setA: Set[int], setB: Set[int]) -> bool:
        'Checks whether or not two sets are equal (contain the same elements).'
        return set(setA) == set(setB)
    
    @staticmethod
    def is_superset_of(setA: Set[int], setB: Set[int]) -> bool:
        'Checks whether or not a set (setA) is a superset of another set (setB).'
        return set(setA) >= set(setB)

        

File: JavaScript/SSSR.js

const Graph = require('./Graph')
class SSSR {
    static getRings(graph, experimental = false) {
        let adjacencyMatrix = graph.getComponentsAdjacencyMatrix();
        if (adjacencyMatrix.length === 0) {
            return null;
        }
        let connectedComponents = Graph.getConnectedComponents(adjacencyMatrix);
        let rings = Array();
        for (var i = 0; i < connectedComponents.length; i++) {
            let connectedComponent = connectedComponents[i];
            let ccAdjacencyMatrix = graph.getSubgraphAdjacencyMatrix([...connectedComponent]);
            let arrBondCount = new Uint16Array(ccAdjacencyMatrix.length);
            let arrRingCount = new Uint16Array(ccAdjacencyMatrix.length);
            for (var j = 0; j < ccAdjacencyMatrix.length; j++) {
                arrRingCount[j] = 0;
                arrBondCount[j] = 0;
                for (var k = 0; k < ccAdjacencyMatrix[j].length; k++) {
                    arrBondCount[j] += ccAdjacencyMatrix[j][k];
                }
            }
            let nEdges = 0;
            for (var j = 0; j < ccAdjacencyMatrix.length; j++) {
                for (var k = j + 1; k < ccAdjacencyMatrix.length; k++) {
                    nEdges += ccAdjacencyMatrix[j][k];
                }
            }
            let nSssr = nEdges - ccAdjacencyMatrix.length + 1;
            let allThree = true;
            for (var j = 0; j < arrBondCount.length; j++) {
                if (arrBondCount[j] !== 3) {
                    allThree = false;
                }
            }
            if (allThree) {
                nSssr = 2.0 + nEdges - ccAdjacencyMatrix.length;
            }
            if (nSssr === 1) {
                rings.push([...connectedComponent]);
                continue;
            }
            if (experimental) {
                nSssr = 999;
            }
            let { d, pe, pe_prime } = SSSR.getPathIncludedDistanceMatrices(ccAdjacencyMatrix);
            let c = SSSR.getRingCandidates(d, pe, pe_prime);
            let sssr = SSSR.getSSSR(c, d, ccAdjacencyMatrix, pe, pe_prime, arrBondCount, arrRingCount, nSssr);
            for (var j = 0; j < sssr.length; j++) {
                let ring = Array(sssr[j].size);
                let index = 0;
                for (let val of sssr[j]) {
                    ring[index++] = connectedComponent[val];
                }
                rings.push(ring);
            }
        }
        return rings;
    }

    static matrixToString(matrix) {
        let str = '';
        for (var i = 0; i < matrix.length; i++) {
            for (var j = 0; j < matrix[i].length; j++) {
                str += matrix[i][j] + ' ';
            }
            str += '\n';
        }
        return str;
    }

    static getPathIncludedDistanceMatrices(adjacencyMatrix) {
        let length = adjacencyMatrix.length;
        let d = Array(length);
        let pe = Array(length);
        let pe_prime = Array(length);
        var l = 0;
        var m = 0;
        var n = 0;
        var i = length;
        while (i--) {
            d[i] = Array(length);
            pe[i] = Array(length);
            pe_prime[i] = Array(length);
            var j = length;
            while (j--) {
                d[i][j] = (i === j || adjacencyMatrix[i][j] === 1) ? adjacencyMatrix[i][j] : Number.POSITIVE_INFINITY;
                if (d[i][j] === 1) {
                    pe[i][j] = [[[i, j]]];
                } else {
                    pe[i][j] = Array();
                }
                pe_prime[i][j] = Array();
            }
        }
        var k = length;
        var j;
        while (k--) {
            i = length;
            while (i--) {
                j = length;
                while (j--) {
                    const previousPathLength = d[i][j];
                    const newPathLength = d[i][k] + d[k][j];
                    if (previousPathLength > newPathLength) {
                        var l, m, n;
                        if (previousPathLength === newPathLength + 1) {
                            pe_prime[i][j] = [pe[i][j].length];
                            l = pe[i][j].length
                            while (l--) {
                                pe_prime[i][j][l] = [pe[i][j][l].length];
                                m = pe[i][j][l].length
                                while (m--) {
                                    pe_prime[i][j][l][m] = [pe[i][j][l][m].length];
                                    n = pe[i][j][l][m].length;
                                    while (n--) {
                                        pe_prime[i][j][l][m][n] = [pe[i][j][l][m][0], pe[i][j][l][m][1]];
                                    }
                                }
                            }
                        } else {
                            pe_prime[i][j] = Array();
                        }
                        d[i][j] = newPathLength;
                        pe[i][j] = [[]];
                        l = pe[i][k][0].length;
                        while (l--) {
                            pe[i][j][0].push(pe[i][k][0][l]);
                        }
                        l = pe[k][j][0].length;
                        while (l--) {
                            pe[i][j][0].push(pe[k][j][0][l]);
                        }
                    } else if (previousPathLength === newPathLength) {
                        if (pe[i][k].length && pe[k][j].length) {
                            var l;
                            if (pe[i][j].length) {
                                let tmp = Array();
                                l = pe[i][k][0].length;
                                while (l--) {
                                    tmp.push(pe[i][k][0][l]);
                                }
                                l = pe[k][j][0].length;
                                while (l--) {
                                    tmp.push(pe[k][j][0][l]);
                                }
                                pe[i][j].push(tmp);
                            } else {
                                let tmp = Array();
                                l = pe[i][k][0].length;
                                while (l--) {
                                    tmp.push(pe[i][k][0][l]);
                                }
                                l = pe[k][j][0].length;
                                while (l--) {
                                    tmp.push(pe[k][j][0][l]);
                                }
                                pe[i][j][0] = tmp
                            }
                        }
                    } else if (previousPathLength === newPathLength - 1) {
                        var l;
                        if (pe_prime[i][j].length) {
                            let tmp = Array();
                            l = pe[i][k][0].length;
                            while (l--) {
                                tmp.push(pe[i][k][0][l]);
                            }
                            l = pe[k][j][0].length;
                            while (l--) {
                                tmp.push(pe[k][j][0][l]);
                            }
                            pe_prime[i][j].push(tmp);
                        } else {
                            let tmp = Array();
                            l = pe[i][k][0].length;
                            while (l--) {
                                tmp.push(pe[i][k][0][l]);
                            }
                            l = pe[k][j][0].length;
                            while (l--) {
                                tmp.push(pe[k][j][0][l]);
                            }
                            pe_prime[i][j][0] = tmp;
                        }
                    }
                }
            }
        }
        return {
            d: d,
            pe: pe,
            pe_prime: pe_prime
        };
    }

    static getRingCandidates(d, pe, pe_prime) {
        let length = d.length;
        let candidates = Array();
        let c = 0;
        for (let i = 0; i < length; i++) {
            for (let j = 0; j < length; j++) {
                if (d[i][j] === 0 || (pe[i][j].length === 1 && pe_prime[i][j] === 0)) {
                    continue;
                } else {
                    if (pe_prime[i][j].length !== 0) {
                        c = 2 * (d[i][j] + 0.5);
                    } else {
                        c = 2 * d[i][j];
                    }
                    if (c !== Infinity) {
                        candidates.push([c, pe[i][j], pe_prime[i][j]]);
                    }
                }
            }
        }
        candidates.sort(function (a, b) {
            return a[0] - b[0];
        });
        return candidates;
    }











    static getSSSR(c, d, adjacencyMatrix, pe, pe_prime, arrBondCount, arrRingCount, nsssr) {
        let cSssr = Array();
        let allBonds = Array();
        for (let i = 0; i < c.length; i++) {
            if (c[i][0] % 2 !== 0) {
                for (let j = 0; j < c[i][2].length; j++) {
                    let bonds = c[i][1][0].concat(c[i][2][j]);
                    for (var k = 0; k < bonds.length; k++) {
                        if (bonds[k][0].constructor === Array) bonds[k] = bonds[k][0];
                    }
                    let atoms = SSSR.bondsToAtoms(bonds);
                    if (SSSR.getBondCount(atoms, adjacencyMatrix) === atoms.size && !SSSR.pathSetsContain(cSssr, atoms, bonds, allBonds, arrBondCount, arrRingCount)) {
                        cSssr.push(atoms);
                        allBonds = allBonds.concat(bonds);
                    }
                    if (cSssr.length > nsssr) {
                        return cSssr;
                    }
                }
            } else {
                for (let j = 0; j < c[i][1].length - 1; j++) {
                    let bonds = c[i][1][j].concat(c[i][1][j + 1]);
                    for (var k = 0; k < bonds.length; k++) {
                        if (bonds[k][0].constructor === Array) bonds[k] = bonds[k][0];
                    }
                    let atoms = SSSR.bondsToAtoms(bonds);
                    if (SSSR.getBondCount(atoms, adjacencyMatrix) === atoms.size && !SSSR.pathSetsContain(cSssr, atoms, bonds, allBonds, arrBondCount, arrRingCount)) {
                        cSssr.push(atoms);
                        allBonds = allBonds.concat(bonds);
                    }
                    if (cSssr.length > nsssr) {
                        return cSssr;
                    }
                }
            }
        }
        return cSssr;
    }








    static getEdgeCount(adjacencyMatrix) {
        let edgeCount = 0;
        let length = adjacencyMatrix.length;
        var i = length - 1;
        while (i--) {
            var j = length;
            while (j--) {
                if (adjacencyMatrix[i][j] === 1) {
                    edgeCount++;
                }
            }
        }
        return edgeCount;
    }


    static getEdgeList(adjacencyMatrix) {
        let length = adjacencyMatrix.length;
        let edgeList = Array();
        var i = length - 1;
        while (i--) {
            var j = length;
            while (j--) {
                if (adjacencyMatrix[i][j] === 1) {
                    edgeList.push([i, j]);
                }
            }
        }
        return edgeList;
    }



    static bondsToAtoms(bonds) {
        let atoms = new Set();
        var i = bonds.length;
        while (i--) {
            atoms.add(bonds[i][0]);
            atoms.add(bonds[i][1]);
        }
        return atoms;
    }

    static getBondCount(atoms, adjacencyMatrix) {
        let count = 0;
        for (let u of atoms) {
            for (let v of atoms) {
                if (u === v) {
                    continue;
                }
                count += adjacencyMatrix[u][v]
            }
        }
        return count / 2;
    }

















    static pathSetsContain(pathSets, pathSet, bonds, allBonds, arrBondCount, arrRingCount) {
        var i = pathSets.length;
        while (i--) {
            if (SSSR.isSupersetOf(pathSet, pathSets[i])) {
                return true;
            }
            if (pathSets[i].size !== pathSet.size) {
                continue;
            }
            if (SSSR.areSetsEqual(pathSets[i], pathSet)) {
                return true;
            }
        }
        let count = 0;
        let allContained = false;
        i = bonds.length;
        while (i--) {
            var j = allBonds.length;
            while (j--) {
                if (bonds[i][0] === allBonds[j][0] && bonds[i][1] === allBonds[j][1] ||
                    bonds[i][1] === allBonds[j][0] && bonds[i][0] === allBonds[j][1]) {
                    count++;
                }
                if (count === bonds.length) {
                    allContained = true;
                }
            }
        }
        let specialCase = false;
        if (allContained) {
            for (let element of pathSet) {
                if (arrRingCount[element] < arrBondCount[element]) {
                    specialCase = true;
                    break;
                }
            }
        }
        if (allContained && !specialCase) {
            return true;
        }
        for (let element of pathSet) {
            arrRingCount[element]++;
        }
        return false;
    }











    static areSetsEqual(setA, setB) {
        if (setA.size !== setB.size) {
            return false;
        }
        for (let element of setA) {
            if (!setB.has(element)) {
                return false;
            }
        }
        return true;
    }



    static isSupersetOf(setA, setB) {
        for (var element of setB) {
            if (!setA.has(element)) {
                return false;
            }
        }
        return true;
    }
}
module.exports = SSSR;